src/qminer/qminer_core.cpp

```diff
@@ -5877,31 +5910,66 @@ void TIndex::IndexTextPos(const int& KeyId, const TStr& TextStr, const uint64& R
-        const uint64 WordId = WordIdV[WordIdN];
-        // check if first time we see the word
-        if (!WordIdPosH.IsKey(WordId)) {
-            WordIdPosH.AddDat(WordId, TQmGixItemPos(RecId));
+    // for each word we want to build the TQmGixItemPos that is already sorted from beginning. For that reason we
+    // iterate through WordIdV in a way to so that all values added to TQmGixItemPos are in increasing order.
+    // Each time an item TQmGixItemPos becomes full, we add it to Gix and create a new item
+    const int Blocks = (int)ceil(WordIdV.Len() / (float) TQmGixItemPos::Modulo);
+    TIntV OffsetV;
+    for (int N = 0; N < Blocks; N++) { OffsetV.Add(N*TQmGixItemPos::Modulo); }
+
+    // key: word id, val: (Val, ItemPos) where Val is last Val added to ItemPos. Used to make sure we don't add duplicates
+    // if the ItemPos becomes full and we add it to Gix
+    THash<TUInt64, TPair<TInt, TQmGixItemPos>> WordIdPosH;
+    const int MaxVal = TInt::GetMn(WordIdV.Len(), TQmGixItemPos::Modulo);
+
+    for (int IndN = 0; IndN < MaxVal; IndN++) {
+        // the value that we want to add to corresponding TQmGixItemPos
+        const int Val = IndN + 1;
+        // go over all full-size blocks
+        for (int BlockInd = 0; BlockInd < Blocks; BlockInd++) {
+            const int WordIdN = OffsetV[BlockInd] + IndN;
+            // when we are processing the last block, it will likely not be full
+            if (WordIdN >= WordIdV.Len()) { continue; }
+            const uint64 WordId = WordIdV[WordIdN];
+            TPair<TInt, TQmGixItemPos>& LastValItemPosPr = WordIdPosH.AddDat(WordId);
+            // if the item with this value was already added ignore it
+            if (LastValItemPosPr.Val1 == Val) { continue; }
+            // remember the last added value - makes sure we don't add the same value multiple times
+            LastValItemPosPr.Val1 = Val;
+            // add the value Val to the TQmGixItemPos. If the object becomes full and we need to save it, we return true
+            const bool IsFull = LastValItemPosPr.Val2.Add(Val);
+            if (IsFull) {
+                // since we used default constructor we have to set the rec id before putting it to gix
+                LastValItemPosPr.Val2.SetRecId(RecId);
+                // add the full item to gix
+                ((*GixPos).*UpdateMethod)(TKeyWord(KeyId, WordId), LastValItemPosPr.Val2);
+                // create a new empty item that can be used to store other indices for the word
+                LastValItemPosPr.Val2 = TQmGixItemPos();
+            }
         }
-        // remember the position in case there is space left
-        TQmGixItemPos& ItemPos = WordIdPosH.GetDat(WordId);
-        if (ItemPos.IsSpace()) { ItemPos.Add(WordIdN); }
     }
-    // add to index
-    for (auto& WordIdPos : WordIdPosH) {
+
+    // add to index the remaining TQmGixItemPos instances that are not completely filled up
```

Comment:
This is exactly what I want to avoid. Why would we create a vector for each of the words in the document and sort it? This is more than N^2, while we can now do it in linear time.